### PR TITLE
 [FLINK-36521][runtime] Introduce TtlAwareSerializer and TtlAwareSerializerSnapshot

### PR DESCRIPTION
**Pull request by xiangyuf**
## What is the purpose of the change

Introduce TtlAwareSerializer and TtlAwareSerializerSnapshot to resolve the compatibility check between ttlSerializer and original serializer

## Brief change log

*(for example:)*

- Add serializer-snapshot and test-data for `TtlAwareSerializerUpgradeTest` 
- Introduce the implementation of `TtlAwareSerializer` and `TtlAwareSerializerSnapshot`
- Introduce the implementation of `TtlAwareSerializerSnapshotWrapper`

## Verifying this change

This change added tests and can be verified as follows:

  - Added Unit Test in `TtlAwareSerializerTest`
  - Added Unit Test in `TtlAwareSerializerTtlMigrationTest`
  - Added Unit Test in `TtlAwareSerializerUpgradeTest`
  - Added Unit Test in `TtlAwareSerializerSnapshotWrapperTest`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: yes
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no

### CODE REVIEW

**Comment by xiangyuf** [comment]
@flinkbot run azure

**Comment by Zakelly** [comment]
@flinkbot run azure

File: flink-runtime/src/test/java/org/apache/flink/runtime/state/ttl/TtlAwareSerializerUpgradeTest.java [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+import org.apache.flink.api.common.typeutils.base.StringSerializer;
+
+import org.assertj.core.api.Condition;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Objects;
+
+/** State migration test for {@link TtlAwareSerializer}. */
+class TtlAwareSerializerUpgradeTest extends TypeSerializerUpgradeTestBase<String, String> {
```
Comment (Zakelly) [inline flink-runtime/src/test/java/org/apache/flink/runtime/state/ttl/TtlAwareSerializerUpgradeTest.java:36]: I'm not sure why this test is introduced. This test aims to test the upgrade of a `TypeSerializer` but there seems no upgrade at all?

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/ttl/TtlAwareSerializer.java [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+import org.apache.flink.api.common.typeutils.base.ListSerializer;
+import org.apache.flink.api.common.typeutils.base.MapSerializer;
+import org.apache.flink.core.memory.DataInputView;
+import org.apache.flink.core.memory.DataOutputView;
+import org.apache.flink.util.function.SupplierWithException;
+
+import java.io.IOException;
+import java.util.Objects;
+
+/** Serializer to wrap a {@link TypeSerializer} with TTL awareness. */
```
Comment (Zakelly) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/ttl/TtlAwareSerializer.java:32]: Could you please describe more about what `TtlAwareSerializer` and `TtlAwareSerializerSnapshot` aims to and how.

**Review by Zakelly — COMMENTED** [review]
Overall LGTM.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/ttl/TtlAwareSerializer.java [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+                        && ((MapSerializer) wrappedTypeSerializer).getValueSerializer()
+                                instanceof TtlAwareSerializer
+                        && ((TtlAwareSerializer)
+                                        ((MapSerializer) wrappedTypeSerializer)
+                                                .getValueSerializer())
+                                .isTtlEnabled();
+        return ttlSerializer || ttlListSerializer || ttlMapSerializer;
+    }
+
+    public static TypeSerializer<?> wrapTtlAwareSerializer(TypeSerializer<?> typeSerializer) {
```
[the remark this answers is not in the document]
Comment (xiangyuf) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/ttl/TtlAwareSerializer.java:176]: @Zakelly For ListSerializer/MapSerializer,  the return type is not TtlAwareSerializer.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/ttl/TtlAwareSerializer.java (hunk 2 above)
Comment (xiangyuf) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/ttl/TtlAwareSerializer.java]: Done, thank you.

**Comment by xiangyuf** [comment]
@Zakelly hi, all comments have been resolved. Would you kindly take a look when you have time?

**Review by Zakelly — COMMENTED** [review]
LGTM and merging

**Pull request closed without merging**